perf/src/recycled_vec.rs

```diff
@@ -0,0 +1,195 @@
+use {
+    crate::recycler::{RecyclerX, Reset},
+    rand::{seq::SliceRandom, Rng},
+    rayon::prelude::*,
+    serde::{Deserialize, Serialize},
+    std::{
+        ops::{Deref, DerefMut, Index, IndexMut},
+        slice::{Iter, SliceIndex},
+        sync::Weak,
+    },
+};
+
+// A vector wrapper where the underlying memory can be
+// page-pinned. Controlled by flags in case user only wants
+// to pin in certain circumstances.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct RecycledVec<T: Default + Clone + Sized> {
+    x: Vec<T>,
+    #[serde(skip)]
+    recycler: Weak<RecyclerX<RecycledVec<T>>>,
+}
+
+impl<T: Default + Clone + Sized> Reset for RecycledVec<T> {
+    fn reset(&mut self) {
+        self.resize(0, T::default());
+    }
+    fn warm(&mut self, size_hint: usize) {
+        self.resize(size_hint, T::default());
```

Comment:
can be more simply as self.x.reserve()

Reply:
No, reserve does not actually allocate backing memory on the OS level.